compiler/rustc_expand/src/expand.rs

```diff
@@ -206,30 +206,35 @@ ast_fragments! {
     }
 }
 
+pub enum SupportsMacroAttrs {
+    None,
+    OuterOnly,
+    All,
+}
+
 impl AstFragmentKind {
     crate fn dummy(self, span: Span) -> AstFragment {
         self.make_from(DummyResult::any(span)).expect("couldn't create a dummy AST fragment")
     }
 
-    /// Fragment supports macro expansion and not just inert attributes, `cfg` and `cfg_attr`.
-    pub fn supports_macro_expansion(self) -> bool {
+    pub fn supports_macro_attrs(self) -> SupportsMacroAttrs {
         match self {
-            AstFragmentKind::OptExpr
-            | AstFragmentKind::Expr
-            | AstFragmentKind::Pat
+            AstFragmentKind::OptExpr | AstFragmentKind::Expr | AstFragmentKind::Stmts => {
```

Comment:
`AstFragmentKind::Stmts` is kind of a pessimization due to item statements, but it shouldn't matter in practice.